File: src/main/java/nz/co/searchwellington/feeds/RssfeedNewsitemService.java
package nz.co.searchwellington.feeds;

import java.util.Date;
import java.util.Iterator;
import java.util.List;

import nz.co.searchwellington.model.Feed;
import nz.co.searchwellington.model.frontend.FrontendFeedNewsitem;
import nz.co.searchwellington.repositories.HibernateResourceDAO;

import org.apache.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class RssfeedNewsitemService {

	private static Logger log = Logger.getLogger(RssfeedNewsitemService.class);

	private CachingRssfeedNewsitemService cachingRssfeedNewsitemService;
	private HibernateResourceDAO resourceDAO;
	
	@Autowired
	public RssfeedNewsitemService(
			CachingRssfeedNewsitemService cachingRssfeedNewsitemService,
			HibernateResourceDAO resourceDAO,
			FeednewsItemToNewsitemService feednewsItemToNewsitemService) {
		this.cachingRssfeedNewsitemService = cachingRssfeedNewsitemService;
		this.resourceDAO = resourceDAO;
	}

	public List<FrontendFeedNewsitem> getFeedNewsitems(Feed feed) {
		return cachingRssfeedNewsitemService.getFeedNewsitems(feed);
	}
	
	public final Date getLatestPublicationDate(Feed feed) {
		Date latestPublicationDate = null;
		List<FrontendFeedNewsitem> feeditems = getFeedNewsitems(feed);
		for (FrontendFeedNewsitem feeditem : feeditems) {
			if (feeditem.getDate() != null && (latestPublicationDate == null || feeditem.getDate().after(latestPublicationDate))) {
				latestPublicationDate = feeditem.getDate();
			}
		}
		return latestPublicationDate;
	}
	
	public FrontendFeedNewsitem getFeedNewsitemByUrl(Feed feed, String url) {
		List<FrontendFeedNewsitem> feedNewsitems = this.getFeedNewsitems(feed);
		Iterator<FrontendFeedNewsitem> i = feedNewsitems.iterator();
		while (i.hasNext()) {
			FrontendFeedNewsitem feedNewsitem = i.next();
			if (feedNewsitem.getUrl() != null && feedNewsitem.getUrl().equals(url)) {
				return feedNewsitem;
			}
		}
		return null;
	}
	
	public boolean isUrlInAcceptedFeeds(String url) {
		log.debug("Looking for url in accepted feeds: " + url);
		for(Feed feed : resourceDAO.getAllFeeds()) {
			if (feed.getAcceptancePolicy().equals("accept") || feed.getAcceptancePolicy().equals("accept_without_dates")) {
				log.debug("Checking feed: " + feed.getName());
				List <FrontendFeedNewsitem> feednewsItems = this.getFeedNewsitems(feed);
				for (FrontendFeedNewsitem feedNewsitem : feednewsItems) {
					log.debug("Checking feeditem: " + feedNewsitem.getUrl());
					if (feedNewsitem.getUrl().equals(url)) {
						log.debug("Found: " + url);
						return true;
					}
				}
			}
		}
		return false;		
	}
	
}

File: src/main/scala/nz/co/searchwellington/feeds/RssfeedNewsitemService.scala
package nz.co.searchwellington.feeds

import java.util.Date

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.feeds.whakaoko.WhakaokoFeedReader
import nz.co.searchwellington.feeds.whakaoko.model.{FeedItem, Subscription}
import nz.co.searchwellington.model.Feed
import nz.co.searchwellington.repositories.mongo.MongoRepository
import org.apache.log4j.Logger
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import scala.concurrent.{ExecutionContext, Future}

@Component class RssfeedNewsitemService @Autowired()(whakaokoFeedReader: WhakaokoFeedReader,
                                                     mongoRepository: MongoRepository)
  extends ReasonableWaits {

  private val log = Logger.getLogger(classOf[FeedReaderRunner])

  def getChannelFeedItems(page: Int, subscriptions: Seq[Subscription])(implicit ec: ExecutionContext): Future[Seq[(FeedItem, Feed)]] = {

    def decorateFeedItemsWithFeeds(feedItems: Seq[FeedItem], subscriptions: Seq[Subscription]): Future[Seq[(FeedItem, Feed)]] = {
      val eventualMaybeFeedsBySubscriptionId = Future.sequence(subscriptions.map { s =>
        mongoRepository.getFeedByUrl(s.url).map { fo =>
          (s.id, fo)
        }
      })

      val eventualFeedsBySubscriptionId = eventualMaybeFeedsBySubscriptionId.map { f =>
        f.flatMap { i =>
          i._2.map { f =>
            (i._1, f)
          }
        }.toMap
      }

      eventualFeedsBySubscriptionId.map { feeds =>
        feedItems.flatMap { fi =>

          if (feeds.get(fi.subscriptionId).isEmpty) {
            log.warn("No local feed found for feed item with subscription id (" + fi.subscriptionId + "): " + fi)
          }

          feeds.get(fi.subscriptionId).map { feed =>
            (fi, feed)
          }
        }
      }
    }

    for {
      channelFeedItems <- whakaokoFeedReader.fetchChannelFeedItems(page)
      channelFeedItemsWithFeeds <- decorateFeedItemsWithFeeds(channelFeedItems, subscriptions)

    } yield {
      channelFeedItemsWithFeeds
    }

  }

  def getFeedItemsAndDetailsFor(feed: Feed)(implicit ec: ExecutionContext): Future[Either[String, ((Seq[FeedItem], Long), Subscription)]] = {
    whakaokoFeedReader.fetchFeedItems(feed)
  }

  def getLatestPublicationDate(feed: Feed)(implicit ec: ExecutionContext): Future[Option[Date]] = {
    getFeedItemsAndDetailsFor(feed).map { r =>
      r.toOption.flatMap { right =>
        latestPublicationDateOf(right._1._1)
      }
    }
  }

  def latestPublicationDateOf(feedItems: Seq[FeedItem]): Option[Date] = {
    val publicationDates = feedItems.flatMap(fi => fi.date.map(_.toDate))
    if (publicationDates.nonEmpty) {
      Some(publicationDates.max)
    } else {
      None
    }
  }

}
